src/addrdb.cpp

```diff
@@ -119,18 +176,54 @@ bool DeserializeFileDB(const fs::path& path, Data& data)
 }
 } // namespace
 
-CBanDB::CBanDB(fs::path ban_list_path) : m_ban_list_path(std::move(ban_list_path))
+CBanDB::CBanDB(fs::path ban_list_path)
+    : m_banlist_dat(ban_list_path.string() + ".dat"),
+      m_banlist_json(ban_list_path.string() + ".json")
 {
 }
 
 bool CBanDB::Write(const banmap_t& banSet)
 {
-    return SerializeFileDB("banlist", m_ban_list_path, banSet);
+    std::vector<std::string> errors;
+    if (util::WriteSettings(m_banlist_json, {{JSON_KEY, BanMapToJson(banSet)}}, errors)) {
+        return true;
```

Comment:
I deliberately avoid removing `banlist.dat` automatically because that would prevent downgrading to an older version (and keeping the ban list).

Something like this may be a good idea:

```cpp
--- i/src/addrdb.cpp
+++ w/src/addrdb.cpp
@@ -185,12 +185,18 @@ CBanDB::CBanDB(fs::path ban_list_path)
 }

 bool CBanDB::Write(const banmap_t& banSet)
 {
     std::vector<std::string> errors;
     if (util::WriteSettings(m_banlist_json, {{JSON_KEY, BanMapToJson(banSet)}}, errors)) {
+        if (fs::exists(m_banlist_dat)) {
+            LogPrintf("Successfully saved %s to disk. %s can be removed manually now if "
+                      "downgrading to an older version is not planned.\n",
+                      m_banlist_json.string(),
+                      m_banlist_dat.string());
+        }
         return true;
     }

     for (const auto& err : errors) {
         error("%s", err);
     }
```

but it would be printed **every** time which would be annoying. There is some doc added to `doc/files.md` which should clarify things if the user wonders why there are two files. Maybe leave it as is?

We can add an automatic removal of `banlist.dat` one major version after this change is released.

Reply:
Removing `banlist.dat` one major version later seems like a fair choice!